File: src/main/java/recipeapplication/security/RecipeUserDetails.java
package recipeapplication.security;

import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import recipeapplication.model.User;

import java.util.Collection;
import java.util.Collections;

public class RecipeUserDetails implements UserDetails
{
    private User user;
    private boolean changePasswordAccess;

    public RecipeUserDetails(User user)
    {
        this.user = user;
        changePasswordAccess = false;
    }

    public Collection<? extends GrantedAuthority> getAuthorities()
    {
        if (!changePasswordAccess)
        {
            return Collections.singletonList(new SimpleGrantedAuthority(Role.USER.toString()));
        }

        return null;
    }

    public String getUsername()
    {
        return user.getUsername();
    }

    public String getPassword()
    {
        return user.getPassword();
    }

    public User getUser()
    {
        return user;
    }

    public boolean isAccountNonExpired()
    {
        return true;
    }

    public boolean isAccountNonLocked()
    {
        return true;
    }

    public boolean isCredentialsNonExpired()
    {
        return true;
    }

    public boolean isEnabled()
    {
        return true;
    }

    public void setChangePasswordAccess(boolean changePasswordAccess)
    {
        this.changePasswordAccess = changePasswordAccess;
    }

    public boolean isChangePasswordAccess()
    {
        return changePasswordAccess;
    }
}

File: src/main/kotlin/thecookingpot/security/RecipeUserDetails.kt
package thecookingpot.security

import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import thecookingpot.model.User

class RecipeUserDetails(val user: User) : UserDetails {
    var changePasswordAccess = false

    override fun getAuthorities(): Collection<GrantedAuthority?> {
        return if (!changePasswordAccess) {
            listOf(SimpleGrantedAuthority(Role.USER.toString()))
        } else emptyList() // TODO I need to validate this is ok
    }

    override fun getUsername(): String {
        return user.username
    }

    override fun getPassword(): String {
        return user.password
    }

    override fun isAccountNonExpired(): Boolean {
        return true
    }

    override fun isAccountNonLocked(): Boolean {
        return true
    }

    override fun isCredentialsNonExpired(): Boolean {
        return true
    }

    override fun isEnabled(): Boolean {
        return true
    }
}